src/Orleans.Runtime/MembershipService/MembershipTableManager.cs

```diff
@@ -93,6 +93,19 @@ public async Task Refresh()
             await pending;
         }
 
+        public async Task RefreshFromSnapshot(MembershipTableSnapshot snapshot)
+        {
+            // Check if a refresh is underway
+            var pending = this.pendingRefresh;
+            if (pending != null && !pending.IsCompleted)
+            {
+                await pending;
+            }
+
+            // If we are behind, let's take directly the snapshot in param
+            this.updates.Publish(snapshot);
```

Comment:
I think this should be TryPublish instead, since the publish may fail if an equal/later version was already published